api/src/main/java/org/eclipse/microprofile/rest/client/RestClientBuilder.java

```diff
@@ -17,13 +17,29 @@
 package org.eclipse.microprofile.rest.client;
 
 import javax.annotation.Priority;
+import javax.ws.rs.core.Configurable;
 import java.net.URL;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 import java.util.ServiceLoader;
+import java.util.function.ToIntFunction;
 
-public abstract class RestClientBuilder {
+/**
+ * <p>
+ * This is the main entry point for creating a Type Safe Rest Client.  Implementations are expected
+ * to implement this class and register a service provider via {@link ServiceLoader} that works within their implementation.
+ * <p>
+ * Invoking <pre>RestClientBuilder.newBuilder()</pre> is intended to always create a new instance, not use a cached version.
+ * <p>
+ * If multiple implementations of RestClientBuilder are discovered, the one with the highest value of the {@link Priority} annotation is used
+ * <p>
+ * The {@link ServiceLoader} will first search via the current Thread's Context ClassLoader, then {@link RestClientBuilder}'s {@link ClassLoader}
+ * <p>
+ * The <pre>RestClientBuilder</pre> is a {@link Configurable} class as defined by JAX-RS.  This allows a user to register providers,
+ * implementation specific configuration.
+ */
+public abstract class RestClientBuilder implements Configurable<RestClientBuilder>{
```

Comment:
This class needs to have setBuilder for it to be worked in OSGi environment, as noted in one of the builder pattern issue.

Reply:
@Emily-Jiang understood, we are tracking your builder pattern concerns in a separate ticket.

Reply:
I confess to not understanding requirements for OSGI.

However, it seems odd that we're having to add methods specifically for OSGI here, but `ClientBuilder` in JAX-RS which is also an abstract class doesn't have such methods.

@Emily-Jiang, how do Java EE 7/8 APIs work in OSGI without similar methods and bnd configuration?

Reply:
I'm curious about this as well, but very willing to support OSGi.

When i look around the Open Liberty code, I found the following of interest:

https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.javaee.jaxrs.2.1/resources/META-INF/services/javax.ws.rs.client.ClientBuilder
https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/JAXRSClientBuilderImpl.java

However, JAX-RS itself is using `FactoryFinder` to locate the service... so I'm wondering if we just use that will it solve the problem? http://grepcode.com/file/repo1.maven.org/maven2/javax.ws.rs/javax.ws.rs-api/2.0/javax/ws/rs/client/FactoryFinder.java?av=f

Reply:
See what MP Config ConfigProviderResolver.setInstance(). The reason is that service loader pattern is not supported in OSGi. OSGi directly uses service discovery to discover the service of ConfigProviderResolver. The same thing is for MicroProfile Health Check. Which JavaEE 7/8 APIs did you see with a service loader pattern? Bnd configuration is to build a proper bundle to be used in OSGi. Some JavaEE 7/8 did not manage to build a proper bundle nor have their version correctly done. We should not follow the wrong path. Instead, MicroProfile releases should work well with OSGi for our future modularity story. 

By the way, in my recent talk of MP Config in OSGi, I was trying to promote MP programming model to be adopted by OSGi Alliance.

Reply:
@Emily-Jiang I'm just trying to understand why OSGI is requiring certain things that are not part of Java EE APIs, as I have zero understanding of OSGI. For instance, the JAX-RS APIs in Java EE 7.

Are you saying that the JAX-RS API doesn't do enough from an OSGI perspective? Are there outstanding issues that haven't been resolved for this?

I'm not trying to block OSGI support. I'm just trying to understand what's actually required, as it seems to be adding things to the /api side that I wouldn't necessarily have expected.

With respect to the OSGI Maven dependency. Can that be marked as "optional" so that it doesn't become a transitive dependency of non OSGI containers?

Reply:
While I understand the need, and agree that we need to continue discussing it, the usage of OSGi is not relevant to this PR.  Please continue the discussion within https://github.com/eclipse/microprofile-rest-client/issues/15